system/include/emscripten/val.h

```diff
@@ -496,10 +496,52 @@ namespace emscripten {
     template<typename T>
     std::vector<T> vecFromJSArray(val v) {
         auto l = v["length"].as<unsigned>();
-
         std::vector<T> rv;
-        for(unsigned i = 0; i < l; ++i) {
-            rv.push_back(v[i].as<T>());
+
+        if (internal::typeSupportsMemoryView<T>()) {
+            rv.resize(l);
+
+            const char *arrayType;
+
+            switch (sizeof(T)) {
+                case 1:
+                    if (std::is_unsigned<T>::value) {
+                        arrayType = "Uint8Array";
+                    } else {
+                        arrayType = "Int8Array";
+                    }
+                    break;
+                case 2:
+                    if (std::is_unsigned<T>::value) {
+                        arrayType = "Uint16Array";
+                    } else {
+                        arrayType = "Int16Array";
+                    }
+                    break;
+                case 4:
+                    if (std::is_unsigned<T>::value) {
+                        arrayType = "Uint32Array";
+                    } else if (std::is_integral<T>::value) {
+                        arrayType = "Int32Array";
+                    } else {
+                        arrayType = "Float32Array";
+                    }
+                    break;
+                case 8:
+                    arrayType = "Float64Array";
+                    break;
+                default:
+                    throw;
```

Comment:
Hmm, I thought that `throw;` is only allowed inside a `catch()` handler to rethrow the exception object. What does a `throw;` statement outside any `catch()` handlers throw? Does this compile?

Reply:
This syntax (used outside a `catch` block) actually terminates the program. I added this because it should never be reached (due to the `typeSupportsMemoryView` check before).

I'll replace it with an explaining comment and throw a `logic_error`.